src/ecdsa/mod.rs

```diff
@@ -0,0 +1,127 @@
+use std::convert::TryInto;
+
+use ecdsa::{
+    elliptic_curve::{sec1::ToEncodedPoint, Field},
+    hazmat::SignPrimitive,
+};
+use message_digest::MessageDigest;
+use serde::{Deserialize, Serialize};
+use tracing::error;
+
+use crate::{
+    crypto_tools::{k256_serde, message_digest, rng},
+    sdk::api::{BytesVec, TofnFatal, TofnResult},
+};
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct KeyPair {
+    signing_key: k256_serde::SecretScalar,
+    verifying_key: k256_serde::ProjectivePoint,
+}
+
+impl KeyPair {
+    /// SEC1-encoded compressed curve point
+    pub fn encoded_verifying_key(&self) -> TofnResult<[u8; 33]> {
+        // TODO make this work with k256_serde::ProjectivePoint::to_bytes
+        self.verifying_key
+            .as_ref()
+            .to_affine()
+            .to_encoded_point(true)
+            .as_bytes()
+            .try_into()
+            .map_err(|_| {
+                error!("failure to convert ecdsa verifying key to 33-byte array");
+                TofnFatal
+            })
+    }
+
+    pub fn signing_key(&self) -> &k256_serde::SecretScalar {
+        &self.signing_key
+    }
+}
+
+pub fn keygen(
+    secret_recovery_key: &rng::SecretRecoveryKey,
```

Comment:
Not too important but I wanna mention it for the future:
Since `tofnd` doesn't get all parties now, there is no `party_uid` to `party_id` conversion, right? Maybe we should still get the `party_uid` from axelar-core to use it for domain separation here (as opposed to `party_id`)?